subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/ArtifactAttributeMatchingCache.java

```diff
@@ -0,0 +1,117 @@
+
+    private static final Transformer<List<File>, File> NO_TRANSFORM = new Transformer<List<File>, File>() {
+        @Override
+        public List<File> transform(File file) {
+            return null;
+        }
+    };
+
+    public ArtifactAttributeMatchingCache(ArtifactTransformRegistrations artifactTransformRegistrations, AttributesSchema schema) {
+        this.attributeMatcher = new ArtifactAttributeMatcher(schema);
+        this.artifactTransformRegistrations = (ArtifactTransformRegistrationsInternal) artifactTransformRegistrations;
+    }
+    boolean areMatchingAttributes(AttributeContainer artifact, AttributeContainer target) {
+        return matchAttributes(artifact, target, false);
+    }
+
+    Transformer<List<File>, File> getTransform(AttributeContainer artifact, AttributeContainer target) {
+        AttributeSpecificCache toCache = getCache(target.getAttributes());
+        Transformer<List<File>, File> transformer = toCache.transforms.get(target);
+        if (transformer == null) {
+            for (ArtifactTransformRegistration transformReg : artifactTransformRegistrations.getTransforms()) {
+                if (matchAttributes(transformReg.from, artifact, true)
+                    && matchAttributes(transformReg.to, target, true)) {
+
+                    transformer = transformReg.getTransform();
+                    toCache.transforms.put(artifact.getAttributes(), transformer == null ? NO_TRANSFORM : transformer);
+                    return transformer;
+                }
+            }
+        }
+        return transformer == NO_TRANSFORM ? null : transformer;
+    }
+
+    private AttributeSpecificCache getCache(AttributeContainer attributes) {
+        AttributeSpecificCache cache = attributeSpecificCache.get(attributes);
+        if (cache == null) {
+            cache = new AttributeSpecificCache();
+            attributeSpecificCache.put(attributes, cache);
+        }
+        return cache;
+    }
+
+    private boolean matchAttributes(AttributeContainer artifact, AttributeContainer target, boolean incompleteCandidate) {
+        Map<AttributeContainer, Boolean> cache;
+        if (incompleteCandidate) {
+            cache = getCache(target).partialMatches;
+        } else {
+            cache = getCache(target).exactMatches;
+        }
+
+        Boolean match = cache.get(artifact);
+        if (match == null) {
+            match = attributeMatcher.attributesMatch(artifact, target, incompleteCandidate);
+            cache.put(artifact, match);
+        }
+        return match;
+    }
+
+    List<ResolvedArtifact> getTransformedArtifacts(ResolvedArtifact artifact, AttributeContainer target) {
+        return getCache(target).transformedArtifacts.get(artifact);
+    }
+
+    void putTransformedArtifact(ResolvedArtifact artifact, AttributeContainer target, List<ResolvedArtifact> transformResults) {
```

Comment:
I think we should replace get/put with a get-method that takes a callback on cache misses. That makes this class much harder to abuse.